server/src/main/java/org/elasticsearch/tracing/Tracer.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.tracing;
+
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.core.Releasable;
+
+import java.util.Map;
+
+/**
+ * Represents a distributed tracing system that keeps track of the start and end
+ * of various activities in the cluster. Traces are composed of "spans", each of
+ * which represents some piece of work with a duration. Spans are nested to
+ * create a parent-child hierarchy.
+ * <p>
+ * You can open a span using
+ * {@link #onTraceStarted(ThreadContext, String, String, Map)}, and stop it using
+ * {@link #onTraceStopped(String)}, at which point the tracing system will queue
+ * the data to be sent somewhere for processing and storage.
+ * <p>
+ * You can add additional data to a span using the {@code setAttribute(...)}
+ * methods, e.g.  {@link #setAttribute(String, String, String)}. This allows you
+ * to attach data that is not available when the span is opened.
+ */
+public interface Tracer {
```

Comment:
I think we should add a comment here that the methods must have performance characteristics similar to logging since they're all used on hot transport threads.
If we implement this interface to send out network requests on any method we need to make sure that it's all non-blocking and with appropriate queuing to avoid slowing down transport threads.